### PR TITLE
feat: Add snippets for all Layer classes [#34]

### PR DESCRIPTION
**Pull request by BochaberiDaisy**
Fix: #34 

@hhkaos, please check the PR.

I have added snippets to all the Layers. 
Let me know if any changes are needed 🙂

### CODE REVIEW

File: snippets/javascript.json [hunk 1]
```diff
@@ -653,5 +653,868 @@
...
+    "body": [
+    "CSVLayer({",
+    "\turl: \"${1:https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/2.5_week.csv}\",",
+    "\tcopyright: \"${2:USGS Earthquakes}\",",
+    "\tpopupTemplate: ${3:template}",
+    "});",
+    "map.add(csvLayer);"
+    ],
+    "description": "Add a CSVLayer with a URL",
+    "prefix": "csvLayer"
```
Comment (hhkaos) [inline snippets/javascript.json:667]:
```suggestion
    "prefix": "CSVLayer"
```
Prefix on **Class constructors** should match the SDK capitalization according to [the conventions](https://github.com/Esri/arcgis-js-vscode-snippets/blob/master/CONTRIBUTING.md#prefix)

File: snippets/javascript.json [hunk 2]
```diff
@@ -653,5 +653,868 @@
...
+    "});",
+    "let url = URL.createObjectURL(blob);",
+    "",
+    "const layer = new CSVLayer({",
+    "\turl: ${2:url}",
+    "});",
+    "map.add(layer);"
+    ],
+    "description": "Create and add a CSVLayer from a Blob object",
+    "prefix": "csvLayerFromBlob"
```
Comment (hhkaos) [inline snippets/javascript.json:682]:
```suggestion
    "prefix": "CSVLayerFromBlob"
```

File: snippets/javascript.json [hunk 3]
```diff
@@ -653,5 +653,868 @@
...
+  "Add GeoJSONLayer with URL": {
+    "body": [
+    "GeoJSONLayer({",
+    "\turl: \"${1:https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/all_month.geojson}\",",
+    "\tcopyright: \"${2:USGS Earthquakes}\"",
+    "});",
+    "map.add(geojsonlayer);"
+    ],
+    "description": "Add GeoJSONLayer using URL",
+    "prefix": "geoJSONLayer"
```
Comment (hhkaos) [inline snippets/javascript.json:693]:
```suggestion
    "prefix": "GeoJSONLayer"
```

File: snippets/javascript.json [hunk 4]
```diff
@@ -653,5 +653,868 @@
...
+    "map.add(geoRSSLayer);"
+    ],
+    "description": "Add GeoRSSLayer using URL",
+    "prefix": "addGeoRSSLayer"
+  },
+  "Add Graphics to GraphicsLayer": {
+    "body": [
+    "map.add(graphicsLayer);",  
+    "Graphic({",
+    "\tgeometry: ${1:point,polyline, polygon,}",
```
Comment (hhkaos) [inline snippets/javascript.json:709]:
```suggestion
    "\tgeometry: ${1:Point,Polyline, Polygon}",
```

File: snippets/javascript.json [hunk 5]
```diff
@@ -653,5 +653,868 @@
...
+  },
+  "Create GeoRSSLayer with URL": {
+    "body": [
+    "GeoRSSLayer({",
+    "\turl: \"${1:https://arcgis.github.io/arcgis-samples-javascript/sample-data/layers-georss/sample-georss.xml}\"",
+    "});",
+    "map.add(geoRSSLayer);"
+    ],
+    "description": "Add GeoRSSLayer using URL",
+    "prefix": "addGeoRSSLayer"
```
Comment (hhkaos) [inline snippets/javascript.json:703]:
```suggestion
    "prefix": "GeoRSSLayer"
```

File: snippets/javascript.json [hunk 6]
```diff
@@ -653,5 +653,868 @@
...
+  },
+  "Add Graphics to GraphicsLayer": {
+    "body": [
+    "map.add(graphicsLayer);",
```
Comment (hhkaos) [inline snippets/javascript.json:707]:
```suggestion
```

File: snippets/javascript.json [hunk 7]
```diff
@@ -653,5 +653,868 @@
...
+  },
+  "Create FeatureLayer with PortalItem": {
+      "body": [
+        "FeatureLayer({",
+        "\tportalItem: {",
+        "\t\tid: \"${1:123f4410054b43d7a0bacc1533ceb8dc}\"",
+        "\t}",
+        "});",
+        "",
+        "table.load().then(function() {",
```
Comment (hhkaos) [inline snippets/javascript.json:725]: I feel this table load could be omited

File: snippets/javascript.json [hunk 8]
```diff
@@ -653,5 +653,868 @@
...
+        "\t\t\talias: \"${2:ObjectID}\",",
+        "\t\t\ttype: \"${3:oid}\"",
+        "\t\t},",
+        "\t\t{",
+        "\t\t\tname: \"${4:type}\",",
+        "\t\t\talias: \"${5:Type}\",",
+        "\t\t\ttype: \"${6:string}\"",
+        "\t\t},",
+        "\t\t{",
+        "\t\t\tname: \"${7:place}\",",
```
Comment (hhkaos) [inline snippets/javascript.json:749]: Probably we can remove this `place` field. It is redundant considering we have other "string field" already..

File: snippets/javascript.json [hunk 9]
```diff
@@ -653,5 +653,868 @@
...
+        "\t\t\talias: \"${11:recordedDate}\",",
+        "\t\t\ttype: \"${12:date}\"",
+        "\t\t}",
+        "\t],",
+        "\tdateFieldsTimeZone: \"${13:America/New_York}\",",
+        "\tobjectIdField: \"${14:ObjectID}\",",
+        "\tgeometryType: \"${15:point}\",",
+        "\tspatialReference: { wkid: ${16:4326} },",
+        "\tsource: graphics,",
+        "\tpopupTemplate: pTemplate,",
```
Comment (hhkaos) [inline snippets/javascript.json:764]:
```suggestion
        "\tpopupTemplate: PopupTemplate,",
```

At some point we will introduce snippets for PopupTemplates too: https://github.com/Esri/arcgis-js-vscode-snippets/issues/64

File: snippets/javascript.json [hunk 10]
```diff
@@ -653,5 +653,868 @@
...
+  "Add CSV Layer": {
+    "body": [
+    "CSVLayer({",
+    "\turl: \"${1:https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/2.5_week.csv}\",",
+    "\tcopyright: \"${2:USGS Earthquakes}\",",
+    "\tpopupTemplate: ${3:template}",
+    "});",
+    "map.add(csvLayer);"
+    ],
+    "description": "Add a CSVLayer with a URL",
```
Comment (hhkaos) [inline snippets/javascript.json:666]:
```suggestion
    "description": "Add a CSVLayer (point layer) based on a CSV file (.csv, .txt) with a URL. CSV is a plain-text file format used to represent tabular data, including geographic point features (latitude, longitude). AMD path: esri/layers/CSVLayer | ESM path: @arcgis/core/layers/CSVLayer.js",
```
Following `description` [conventions](https://github.com/Esri/arcgis-js-vscode-snippets/blob/master/CONTRIBUTING.md#description)

File: snippets/javascript.json [hunk 11]
```diff
@@ -653,5 +653,868 @@
...
+    "\ttype: \"plain/text\"",
+    "});",
+    "let url = URL.createObjectURL(blob);",
+    "",
+    "const layer = new CSVLayer({",
+    "\turl: ${2:url}",
+    "});",
+    "map.add(layer);"
+    ],
+    "description": "Create and add a CSVLayer from a Blob object",
```
Comment (hhkaos) [inline snippets/javascript.json:681]:
```suggestion
    "description": "Add a CSVLayer (point layer) based on a CSV file (.csv, .txt) from in-memory CSV data by passing a Blob object. AMD path: esri/layers/CSVLayer | ESM path: @arcgis/core/layers/CSVLayer.js",
```

**Review by hhkaos — CHANGES_REQUESTED** [review]
Hi @BochaberiDaisy ! Thank you very much for the PR. It is a great start and a lot of work, I really appreciate it. I have started checking and I have just noticed we need to review to meet the conventions. A few example I noticed so far:
* Prefixes in class constructors should match the Class Capitalization (e.g. `graphicsLayer` should be GraphicsLayer`; `addGeoRSSLayer` should be `GeoRSSLayer`, etc.). Prefixes starting with lowercase are used for class properties objects, common procedures, and enumerations. If you want to suggest a change, please check the issue #10.
* In the body I have seen inconsistencies with map.add(layerVariable) and not storing the variable. I don't have an strong opinion about having the statement or not, but if we add it, the layer variable should be instantiated within the snippet. If you want to suggest a change, please check the issue #19. 
* The descriptions are not following the conventions. Some descriptions are a little bit short. You can reuse text from the API reference. And the class initiatilization do not finish with the paths. 

I have added some comments just to illustrate some examples. 
Thanks again!

**Comment by BochaberiDaisy** [comment]
Here is the task list for the progress on the changes requested.

> * Vector layers
>   
>   * [x]  [CSVLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-CSVLayer.html#delimiter)
>   * [x]  [GeoJSONLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-GeoJSONLayer.html)
>   * [x]  [GeoRSSLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-GeoRSSLayer.html) (OGC)
>   * [x]  [GraphicsLayer](https://developers.arcgis.com/javascript/latest/add-a-point-line-and-polygon/)
>   * [ ]  [FeatureLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-featurelayer/)
>   * [ ]  [FeatureLayer (client-side)](https://developers.arcgis.com/javascript/latest/layers-and-data/#client-side-data-sources)
>   * [ ]  [MapNotesLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-mapnoteslayer/)
>   * [x]  [OGCFeatureLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-ogcfeaturelayer/) OGC API - Features
>   * [ ]  [RouteLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-RouteLayer.html)
>   * [ ]  [StreamLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-streamlayer/)
>   * [ ]  [VectorTileLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-vectortilelayer/)
>   * [ ]  [WFSLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-wfs/) OGC Web Feature Service (WFS)
> * Raster Layers
>   
>   * [ ]  [ImageryTileLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-imagerylayer/)
>   * [ ]  [ImageryLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-imagerylayer/)
>   * [ ]  [ImageryTileLayer /w COG](https://developers.arcgis.com/javascript/latest/sample-code/layers-imagerytilelayer-cog/)
>   * [ ]  [ImageryTileLayer (Multidimensional)](https://developers.arcgis.com/javascript/latest/sample-code/layers-hosted-imagerytilelayer/)
>   * [ ]  [WebTileLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-WebTileLayer.html) - (XYZ layer / Slippy maps)
>   * [ ]  [WMSLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-wms/) OGC Web Map Service (WMS)
>   * [ ]  [WMTSLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-wmts/) OGC Web Map Tile Service (WMTS)
>   * [ ]  [WCSLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-WCSLayer.html) OGC Web Coverage Service (WCS)
>   * [ ]  [MapImageLayer (Time-based)](https://developers.arcgis.com/javascript/latest/sample-code/time-layer/) Traffic layer
>   * [ ]  [BingMapsLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-BingMapsLayer.html)
>   * [ ]  [OpenStreetMapLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-osm-3d/)
> * 3D layers
>   
>   * [ ]  [SceneLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-SceneLayer.html)
>   * [ ]  [Elevation](https://developers.arcgis.com/javascript/latest/sample-code/scene-elevationlayer/)
>   * [ ]  [PointCloud](https://developers.arcgis.com/javascript/latest/sample-code/layers-pointcloud-portal/)
>   * [ ]  [BuildingSceneLayer](https://developers.arcgis.com/javascript/latest/sample-code/building-scene-layer-slice/)
>   * [ ]  [LineOfSightLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-LineOfSightLayer.html)
>   * [ ]  [DimensionLayer](https://developers.arcgis.com/javascript/latest/sample-code/layers-dimension/)
>   * [ ]  [IntegratedMeshLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-IntegratedMeshLayer.html)
>   * [ ]  [VoxelLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-VoxelLayer.html)
> * Others
>   
>   * [ ]  [MediaLayers](https://developers.arcgis.com/javascript/latest/sample-code/layers-medialayer-control-points/) (Overlayed images with control points)
>   * [ ]  [MediaLayers (/w video)](https://developers.arcgis.com/javascript/latest/sample-code/layers-medialayer-video/)
>   * [ ]  [KML Layer](https://developers.arcgis.com/javascript/latest/sample-code/layers-kml/) (OGC)
>   * [ ]  [GroupLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-GroupLayer.html)
>   * [ ]  [SubtypeGroupLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-SubtypeGroupLayer.html)
>   * [ ]  [TileLayer](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-ImageryTileLayer.html)
>   * [ ]  [KnowledgeGraphLayer](https://developers.arcgis.com/javascript/latest/sample-code/knowledgegraph-knowledgegraphlayer/)

File: snippets/javascript.json [hunk 12]
```diff
@@ -653,5 +653,598 @@
...
+        "\t\t\talias: \"${2:ObjectID}\",",
+        "\t\t\ttype: \"${3:oid}\"",
+        "\t\t},",
+        "\t\t{",
+        "\t\t\tname: \"${4:type}\",",
+        "\t\t\talias: \"${5:Type}\",",
+        "\t\t\ttype: \"${6:string}\"",
+        "\t\t},",
+        "\t\t{",
+        "\t\t\tname: \"${10:recordedDate}\",",
```
Comment (hhkaos) [inline snippets/javascript.json:731]: The order should be updated. It's jumping from 6 to 10

File: snippets/javascript.json [hunk 13]
```diff
@@ -653,5 +653,598 @@
...
+        "\tgeometryType: \"${15:point}\",",
+        "\tspatialReference: { wkid: ${16:4326} },",
+        "\tsource: graphics,",
+        "\trenderer: uvRenderer",
+        "});"
+      ],
+      "description": "Add a FeatureLayer with Client-Side Graphics. If working with spatial layers, the geometry type of the features must be indicated.  Both spatial and non-spatial feature collections require an objectId field.  AMD path: esri/layers/FeatureLayer | ESM path: @arcgis/core/layers/FeatureLayer.js",
+      "prefix": "FeatureLayerwithClientSideGraphics"
+    },
+  "Load WebMap with MapNotes Layer": {
```
Comment (hhkaos) [inline snippets/javascript.json:747]:
```suggestion
  "Load MapNotes Layer": {
```

File: snippets/javascript.json [hunk 14]
```diff
@@ -653,5 +653,598 @@
...
+    "body": [
+    "const mapNotesLayer = new MapNotesLayer({",
```
Comment (hhkaos) [inline snippets/javascript.json:749]:
```suggestion
    "new MapNotesLayer({",
```

File: snippets/javascript.json [hunk 15]
```diff
@@ -653,5 +653,598 @@
...
+    "\tpolylineBarriers: [",
+    "\t\t{ geometry: { paths: [[[${7:-117.3}, ${8:34.05}], [${9:-117.5}, ${10:34.1}]]] } }",
+    "\t],",
+    "\tpolygonBarriers: [",
+    "\t\t{ geometry: { rings: [[[${11:-117.3}, ${12:34.05}], [${13:-117.5}, ${14:34.1}], [${15:-117.4}, ${16:34.2}], [${11}, ${12}]]] } }",
+    "\t]",
+    "});"
+      ],
+      "description": "Add a RouteLayer and visualize the route using stops, point, polygon and polyline barriers. The results of a solved route include overall travel time, distance, and turn-by-turn directions. AMD path: esri/layers/RouteLayer | ESM path: @arcgis/core/layers/RouteLayer.js",
+      "prefix": "routeLayer"
```
Comment (hhkaos) [inline snippets/javascript.json:785]:
```suggestion
      "prefix": "RouteLayer"
```

File: snippets/javascript.json [hunk 16]
```diff
@@ -653,5 +653,598 @@
...
+    "\tgeometryType: \"point\",",
+    "\tmaxReconnectionAttempts: ${3:100},",
+    "\tmaxReconnectionInterval: ${4:10},",
+    "\trenderer: ${5:renderer}",
+    "});"
+    ],
+    "description": "Add a StreamLayer by referencing a custom stream service. The geometryType property is used to set the features' geometry type. AMD path: esri/layers/StreamLayer | ESM path: @arcgis/core/layers/StreamLayer.js",
+    "prefix": "StreamLayerwithWebSocketURL"
+  },
+  "Add Vector Tile Layer": {
```
Comment (hhkaos) [inline snippets/javascript.json:830]: Constructing the vector tile using the style object is OK, but pretty unusual.  I would add [the usual one](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-VectorTileLayer.html#esri/layers/VectorTileLayer):

```js
new VectorTileLayer({
  url: "https://www.arcgis.com/sharing/rest/content/items/4cf7e1fb9f254dcda9c8fbadb15cf0f8/resources/styles/root.json"
});
```

And maybe:
```js
const layer = new VectorTileLayer({
 portalItem:{
   id: "1c365daf37a744fbad748b67aa69dac8"
 }
});
```

File: snippets/javascript.json [hunk 17]
```diff
@@ -653,5 +653,598 @@
...
+        "});"
+      ],
+      "description": "Create a WebTileLayer with a service URL. WebTileLayer provides a simple way to add non-ArcGIS Server map tiles as a layer to a map. The constructor takes a URL template that usually follows a pattern of http://some.domain.com/{level}/{col}/{row}/ where level corresponds to a zoom level, and column and row represent tile column and row, respectively. AMD path: esri/layers/WebTileLayer | ESM path: @arcgis/core/layers/WebTileLayer.js",
+      "prefix": "WebTileLayer"
+    },
+  "Add WMS Layer": {
+    "body": [
+    "new WMSLayer({",
+    "\turl: \"${1:service url}\"",
+    "});"
```
Comment (hhkaos) [inline snippets/javascript.json:924]:
```suggestion
    "});",
    "await wmsLayer.load();"
```
If I remember correctly, in this case, calling load() is important.
Understanding the loadable [pattern](https://developers.arcgis.com/javascript/latest/loadable/)

File: snippets/javascript.json [hunk 18]
```diff
@@ -653,5 +653,598 @@
...
+    ],
+    "description": "Add a MapImageLayer within a specified time extent with URL. MapImageLayer allows you to display and analyze data from sublayers defined in a map service, exporting images instead of features. AMD path: esri/layers/MapImageLayer | ESM path: @arcgis/core/layers/MapImageLayer.js",
+    "prefix": "MapImageLayer"
+  },
+  "OpenStreetMapLayer": {
+      "body": [
+        "new OpenStreetMapLayer();"
+      ],
+      "description": "Add a OpenStreetMapLayer to the map. OpenStreetMapLayer allows the use basemaps from OpenStreetMap. AMD path: esri/layers/OpenStreetMapLayer | ESM path: @arcgis/core/layers/OpenStreetMapLayer.js",
```
Comment (hhkaos) [inline snippets/javascript.json:971]:
```suggestion
      "description": "Add an OpenStreetMapLayer to the map served. Tiles are served from tile.openstreetmap.org. OpenStreetMapLayer allows the use basemaps from OpenStreetMap. AMD path: esri/layers/OpenStreetMapLayer | ESM path: @arcgis/core/layers/OpenStreetMapLayer.js",
```

**Review by hhkaos — CHANGES_REQUESTED** [review]
Amazing work @BochaberiDaisy ! 👏 😍

I just added a few minor things. And I would say just make sure to prettify (fix all indentations).

**Review by hhkaos — CHANGES_REQUESTED** [review]
Amazing work @BochaberiDaisy ! 👏 😍

I just added a few minor things. And I would say just make sure to prettify (fix all indentations).

**Comment by BochaberiDaisy** [comment]
@hhkaos, I have pushed the requested changes. Please check it out.

Thank you for the feedback!

**Review by hhkaos — APPROVED** [review]
Awesome work @BochaberiDaisy , it looks good to me, thank you very much for the hard work!

**Pull request merged**